File: haskell/fizzbuzz.hs
fizzbuzz' :: [(Integer, String)] -> Integer -> String
fizzbuzz' ss n = foldl (\str (num, subst) -> if n `mod` num == 0 then str ++ subst else str ++ "") "" ss

fizzbuzz :: [(Integer, String)] -> Integer -> String
fizzbuzz ss n = if null str then show n else str
  where str = fizzbuzz' ss n

-- mapM_ putStrLn $ map (fizzbuzz [(3, "fizz"), (5, "buzz"), (7, "dazz")]) [1..21]

File: scala/fizzbuzz.scala
import collection.immutable.HashMap
import collection.immutable.Range.Inclusive

def fizz_buzz(r: Inclusive, subs: HashMap[Int, String]): IndexedSeq[String] = {
  r map { x =>
    val s = subs.foldLeft("") { case (acc, (k,v)) => if (x % k == 0) v + acc else acc }
    if (s.isEmpty) x.toString else s
  }
}

fizz_buzz(new Inclusive(14,21,1), HashMap(3 -> "Fizz", 5 -> "Buzz", 7 -> "Dazz")) foreach println
